### PR TITLE
Add pixeldrain backend

### PR DESCRIPTION
**Pull request by Fornax96**
#### What is the purpose of this change?

This change adds support for pixeldrain's experimental filesystem feature. The filesystem is still in closed beta, but I would like to release it with rclone support available. I also use rclone for testing the filesystem itself, it would be useful to have built in support so I don't have to maintain my own fork.

Pixeldrain's filesystem is currently paid-only, but I would be happy to provide some developers with a premium account for testing purposes. [Register an account](https://pixeldrain.com/register) and post your username here, I will upgrade it.

#### Was the change discussed in an issue or in the forum before?

There is a thread asking about pixeldrain support from a while ago: https://forum.rclone.org/t/pixeldrain-support/30124. Though this is about pixeldrain's file sharing component, which this backend does not support.

I also have a thread with questions about the backend implementation: https://forum.rclone.org/t/testsyncwithtrackrenames-failing-with-new-pixeldrain-backend/43044

#### Checklist

- [x] I have read the [contribution guidelines](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#submitting-a-new-feature-or-bug-fix).
- [x] I have added tests for all changes in this PR if appropriate.
- [x] I have added documentation for the changes if appropriate.
- [x] All commit messages are in [house style](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#commit-messages).
- [x] I'm done, this Pull Request is ready for review :-)

image

### CODE REVIEW

**Comment by Fornax96** [comment]
I just fixed the last linter warning. The remaining failures seem to have to do with the S3 backend. So not related to this commit.

**Comment by Fornax96** [comment]
Those linter errors are not in my code.

**Comment by Fornax96** [comment]
Update: Pixeldrain's filesystem is now in open beta: https://pixeldrain.com/filesystem

I periodically merge all updates and re-run the tests. All tests still pass.

**Comment by Fornax96** [comment]
My offer still stands, any rclone developer can post their pixeldrain username and I'll upgrade their account to a paid plan free of charge.

**Comment by DTRHnet** [comment]
Lol, seriously curious why they haven't made a move on your offer.

**Comment by ncw** [comment]
Sorry @Fornax96 been very busy!

I've registered an account as `ncw` (email nick@craig-wood.com )

I'll try to get this merged for 1.66 since it looks ready to go. Would you agree? I haven't reviewed yet, but hopefully will have time tomorrow.

**Comment by Fornax96** [comment]
image

@ncw I have added support for updating metadata through the put and rename endpoints. The Put function now only needs one request. The race conditions are also gone now. All tests pass without errors.

There is one little thing that confuses me. In the Put function (pixeldrain.go:228) I need to manually get the metadata from the context and add the modtime to it. If I don't do this all the tests fail.

But in the Move function (pixeldrain.go:329) I can use `fs.GetConfig(ctx).MetadataSet` and it seems to work fine. Why does this not work for `Put`? I also tried merging `GetMetadataOptions` and `Metadataset` but that doesn't seem to work either.

Is this the intended way to implement metadata in `Put` or is there an easier way?

**Comment by Fornax96** [comment]
I just remembered that I made this.. 

@ncw It's been about half a year since I submitted this PR. Is there any intention of merging it?

**Comment by skapytek** [comment]
Guys im waiting to see pixeldrain support in rclone.

**Comment by skapytek** [comment]
When will rclone support it?

**Comment by garymoon** [comment]
Thanks for this @Fornax96, really tidy PR :blue_heart:

**Comment by ncw** [comment]
@Fornax96 - sorry this PR fell off my radar.

Can you rebase it on master please? I'll try to get it merged for 1.68

**Comment by Fornax96** [comment]
@ncw I see a lot of the test have changed. It will take some effort to get everything passing again. I'll start working on it

File: docs/content/docs.md
Comment (ncw) [inline docs/content/docs.md:283]: Can you drop the whitespace changes from this file please - it makes it very difficult to see what you changed - thanks

File: docs/content/pixeldrain.md
Comment (ncw) [inline docs/content/pixeldrain.md:4]: This will need to be v1.68 (sorry!)

File: backend/pixeldrain/pixeldrain.go
Comment (ncw) [inline backend/pixeldrain/pixeldrain.go:39]: Note that the first line of these `Help:` lines is used as the help for the command line flag.

I think these are easier to read on multiple lines, so I think this should be

```go
Help: `Root of the filesystem to use.

Set to "me" to use your personal filesystem. Set to a shared directory ID to use a shared directory.`,
```

File: backend/pixeldrain/pixeldrain.go
Comment (ncw) [inline backend/pixeldrain/pixeldrain.go:38]: Note that this is called `root_folder_id` in many other backends. Can we call it that?

```
      --box-root-folder-id string          Fill in for rclone to use a non root folder as its starting point
      --drive-root-folder-id string        ID of the root folder
      --filefabric-root-folder-id string   ID of the root folder
      --onedrive-root-folder-id string     ID of the root folder
      --pcloud-root-folder-id string       Fill in for rclone to use a non root folder as its starting point (default "d0")
      --pikpak-root-folder-id string       ID of the root folder
      --sharefile-root-folder-id string    ID of the root folder
```

File: backend/pixeldrain/pixeldrain.go
Comment (ncw) [inline backend/pixeldrain/pixeldrain.go:134]: `> 0` ?

File: backend/pixeldrain/pixeldrain.go
Comment (ncw) [inline backend/pixeldrain/pixeldrain.go:145]: This should use `f` instead of `nil` so that the name of the Fs goes in the logs.

File: backend/pixeldrain/pixeldrain.go
Comment (ncw) [inline backend/pixeldrain/pixeldrain.go:176]: Please put this type assertion at the end like the other backends. There are over 50 backends so consistency is really important for maintenance - thank you :-)

I won't comment on any more of these.

File: backend/pixeldrain/pixeldrain.go
Comment (ncw) [inline backend/pixeldrain/pixeldrain.go:173]: Can you remove the `========` comments please so we can get this into `rclone style` - thanks!

File: backend/pixeldrain/pixeldrain.go
Comment (ncw) [inline backend/pixeldrain/pixeldrain.go:480]: Is 1PB the actual limit? If not then leave this unset and rclone will assume it is unlimited. It then puts OS specific limits on `rclone mount` which will be better than just writing 1PB here.

By unset I mean return `Total:` as `nil` in the `fs.Usage`.

**Review by ncw — COMMENTED** [review]
The backend code is looking very nice and the integration tests run very well - great work :-)

I put some comments inline but they are all minor.

Thank you

**Comment by Fornax96** [comment]
Hold on.. I think I messed up the rebase

**Comment by samrth012** [comment]
> Hold on.. I think I messed up the rebase
> 
> EDIT: Nevermind, it looks alright now. I have addressed all the comments in the last commit.

Does it work for prepaid accounts as well ?
I have been receiving 
`2024/08/04 08:13:07 Failed to create file system for "pixel:": failed to get user data: Get "https://pixeldrain.com/api/user": dial tcp 13.127.247.216:443: i/o timeout`
on my prepaid account for pixeldrain

**Comment by samrth012** [comment]
> > > Hold on.. I think I messed up the rebase
> > > EDIT: Nevermind, it looks alright now. I have addressed all the comments in the last commit.
> > 
> > 
> > I compiled from your fork - Does it work for prepaid accounts as well ? I have been receiving `2024/08/04 08:13:07 Failed to create file system for "pixel:": failed to get user data: Get "https://pixeldrain.com/api/user": dial tcp 13.127.247.216:443: i/o timeout` on my prepaid account for pixeldrain
> 
> Enable the prepaid tier of plan to enable filesystems

Yeah done thanks

**Comment by Fornax96** [comment]
> > Hold on.. I think I messed up the rebase
> > EDIT: Nevermind, it looks alright now. I have addressed all the comments in the last commit.
> 
> I compiled from your fork - Does it work for prepaid accounts as well ? I have been receiving `2024/08/04 08:13:07 Failed to create file system for "pixel:": failed to get user data: Get "https://pixeldrain.com/api/user": dial tcp 13.127.247.216:443: i/o timeout` on my prepaid account for pixeldrain

That IP address does not even belong to pixeldrain. Your DNS server is returning invalid responses.

**Comment by samrth012** [comment]
> > > Hold on.. I think I messed up the rebase
> > > EDIT: Nevermind, it looks alright now. I have addressed all the comments in the last commit.
> > 
> > 
> > I compiled from your fork - Does it work for prepaid accounts as well ? I have been receiving `2024/08/04 08:13:07 Failed to create file system for "pixel:": failed to get user data: Get "https://pixeldrain.com/api/user": dial tcp 13.127.247.216:443: i/o timeout` on my prepaid account for pixeldrain
> 
> That IP address does not even belong to pixeldrain. Your DNS server is returning invalid responses.

Yeah airtel was blocking it, also filesystem was not enabled. Have it working fine, thanks

**Comment by ncw** [comment]
@Fornax96 how are you doing with the changes? Please ping me when you want another review - thanks

**Comment by ncw** [comment]
@Fornax96 thanks. I want to get this merged for v1.68 but I've got too much going on (as usual) so if there is a long pause with me not doing stuff then you have permission to nag me until I've got it merged!

I'll try to review again tomorrow.

File: backend/pixeldrain/api_client.go
Comment (ncw) [inline backend/pixeldrain/api_client.go:237]: It is probably preferable to put `NoResponse: true` in the `rest.Opts` rather than explicitly close the body then it definitely gets closed and it will be consistent with the other calls.

**Review by ncw — COMMENTED** [review]
This is looking very nice now :-)

I put a small number of comments inline.

The backend implements a good number of optional features which is great :-)

Did you consider implementing setting directory modification times? This would involve one more method `DirSetModTime` and working out if you need to set `features.DirModTimeUpdatesOnWrite` or not. If you can't do it that is fine - it is no big loss. Eg 17c0ecc72cd9c2f6eea5a296577607798661e67d and 1693d7ad0f53a550efda8f9608da282dd50aa94a from the SFTP backend

The one thing definitely missing from this backend is use of the pacer. If you looked at your example backend (box say) then you'll see all the `lib/rest` calls are wrapped in pacer retries. This has three main purposes

1. To retry failures on internet errors (eg timeout, connection failed) or on server errors (eg 423 or 500). This is what implements the --low-level-retries and without this your backend will be a lot less robust
2. It implements pacing of calls to the server so it doesn't (for example) send calls more often than every 10mS. Rclone can be very agressive and Go's concurrent nature makes sending 1000s of requests a second very easy (eg user sets `--checkers 1000 --transfers 1000`)
3. In error conditions it implements exponential backoff which is really important for server health.

So I'd recommend you wrap all of the http calls, eg:

```go
	resp, err := f.srv.Call(ctx, &rest.Opts{
        // ...
	})
```

With the pacer, eg (from the box backend)

```go
        var item api.Item
        err = f.pacer.Call(func() (bool, error) {
                resp, err := f.srv.CallJSON(ctx, &opts, nil, &item)
                return shouldRetry(ctx, resp, err)
        })
```

You should wrap the call in `fs.Update` with `pacer.CallNoRetry` which at least implements 2 and 3 above even if it can't retry.

The parameters of the pacer should be chosen so that rclone plays nice with the pixeldrain servers.

With a few small tweaks I see no reason why this backend shouldn't go into v1.68 - thank you for your hard work :-)

File: backend/pixeldrain/pixeldrain.go
Comment (Fornax96) [reply backend/pixeldrain/pixeldrain.go:335]: During implementation it helped me a lot to have these comments so I could easily look up which interfaces corresponded to which functions. I'll remove them in the next commit

File: backend/pixeldrain/api_client.go
Comment (Fornax96) [reply backend/pixeldrain/api_client.go:165]: What are these IDs used for? Pixeldrain has IDs internally but they are not exposed through the API. The API only uses paths for everything.

Directories don't have a size on pixeldrain. And the number of items can only be requested with a new stat request on the dir itself, so that's not really worth it.

File: backend/pixeldrain/api_client.go
Comment (Fornax96) [reply backend/pixeldrain/api_client.go:237]: Wouldn't the response get discarded if I set NoReponse here?

File: docs/content/pixeldrain.md
Comment (Fornax96) [reply docs/content/pixeldrain.md:98]: Good tip, I'll add it to the docs

File: backend/pixeldrain/api_client.go
Comment (ncw) [reply backend/pixeldrain/api_client.go:165]: The IDs here don't get used by rclone but can be show to the user so if you think the user might like to see them, then put them in (is that the ID on the sharing directory maybe?)

There might be nothing you can sensibly set here and that is fine, I just wanted to check we didn't miss anything.

File: backend/pixeldrain/api_client.go
Comment (ncw) [reply backend/pixeldrain/api_client.go:237]: Looks like I mis-read the code!

`CallJSON` with a non-nil response shouldn't need to have the body closed so I think you can remove the `resp.Body.Close()`.

https://github.com/rclone/rclone/blob/e334366345440af8c8392d47e662a8bc3607a4a4/lib/rest/rest.go#L449-L450

File: backend/pixeldrain/api_client.go
Comment (Fornax96) [reply backend/pixeldrain/api_client.go:165]: Ah, I get it now. I've added the public ID for shared directories

**Comment by Fornax96** [comment]
@ncw 

My last commit added the DirSetModTime method, added the API pacer and removed Body.Close calls on CallJSON requests. Pixeldrain does not really have an rate limit, but I added support for the 429 error code in case I ever add it.

**Comment by Fornax96** [comment]
@ncw Fixed the linter error

File: backend/pixeldrain/api_client.go
Comment (ncw) [inline backend/pixeldrain/api_client.go:133]: This is a neat idea (calling this function with the output of f.src.CallJSON), but I don't think it quite works as the retry function needs to check the context too so I think you'll need to use the standard `shouldRetry` function.

Here is a `shouldRetry` function which should work for you - feel free to adjust the error codes as appropriate. Note that `fserrors.ShouldRetry(err)` is a much more powerful version of `os.IsTimeout` and note also the `fserrors.ContextError(ctx, &err)` which is important as rclone sometimes cancels requests and we don't want to retry them.

```go
// retryErrorCodes is a slice of error codes that we will retry
var retryErrorCodes = []int{
	429, // Too Many Requests.
	500, // Internal Server Error
	502, // Bad Gateway
	503, // Service Unavailable
	504, // Gateway Timeout
}

// shouldRetry returns a boolean as to whether this resp and err
// deserve to be retried.  It returns the err as a convenience
func shouldRetry(ctx context.Context, resp *http.Response, err error) (bool, error) {
	if fserrors.ContextError(ctx, &err) {
		return false, err
	}
	return fserrors.ShouldRetry(err) || fserrors.ShouldRetryHTTP(resp, retryErrorCodes), err
}
```

**Review by ncw — COMMENTED** [review]
Very good :-)

I think we need to change the retry function a bit but this is nearly there.

Thank you.

File: backend/pixeldrain/api_client.go
Comment (Fornax96) [reply backend/pixeldrain/api_client.go]: I think think this version will work fine without modification. I've added it in the last commit

**Review by ncw — APPROVED** [review]
That looks great - thank you :-)

I ran some smoke tests on the backend and all seems well, so I am going to merge now.

If you find problems please send a PR.

Thank you :-)

**Comment by ncw** [comment]
PS Is the  pixeldrain filesystem publically available yet?

**Comment by ncw** [comment]
PPS I've put pixeldrain on the daily integration tester which you can see the results of here: https://pub.rclone.org/integration-tests/current/

**Comment by Fornax96** [comment]
> PS Is the pixeldrain filesystem publically available yet?

Yes, it's public. Currently it still requires a subscription to use it. I might introduce a free tier later.

**Pull request merged**
